src/main/java/com/svydovets/bibirnate/cache/Cache.java

```diff
@@ -0,0 +1,240 @@
+package com.svydovets.bibirnate.cache;
+
+
+import static com.svydovets.bibirnate.cache.constant.CacheConstant.PARAMETER_CANNOT_BE_NULL;
+
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import com.svydovets.bibirnate.cache.command.extractor.KeyExtractorCommand;
+import com.svydovets.bibirnate.cache.command.extractor.impl.EntityKeyExtractorCommand;
+import com.svydovets.bibirnate.cache.command.extractor.impl.QueryKeyExtractorCommand;
+import com.svydovets.bibirnate.cache.command.invalidation.InvalidationCommand;
+import com.svydovets.bibirnate.cache.command.invalidation.impl.EntityKeyInvalidationCommand;
+import com.svydovets.bibirnate.cache.command.invalidation.impl.QueryKeyInvalidationCommand;
+import com.svydovets.bibirnate.cache.key.Key;
+import com.svydovets.bibirnate.cache.key.parameters.AbstractKeyParam;
+import com.svydovets.bibirnate.exceptions.CacheOverloadException;
+
+/**
+ * This class is LRU cache realization.
+ * LRU builds on the {@link Key#getUpdated()} field. Each time when we extract a {@link Key} inside the
+ * {@link KeyExtractorCommand} realisation we do {@link Key#update()}.
+ * For controlling cache state we have a mechanism that makes autoClean in the
+ * {@link Cache#put(AbstractKeyParam, Object)}.
+ * Cache is constructed on the {@link ConcurrentHashMap}. This map consists from the {@link Key} (that is the custom
+ * class) and {@link Object} as value (we are able to put an entity, collection or other types of objects).
+ * For extraction from the cache is the method {@link Cache#get(AbstractKeyParam, Class)} where the second parameter is
+ * the realization of the {@link KeyExtractorCommand}. This way was chosen for the case if we provide a new extension
+ * of the {@link AbstractKeyParam}, then we can provide a new command for extraction and add it into cache via
+ * {@link Cache#addKeyExtractorCommand(KeyExtractorCommand)}.
+ * For invalidation the same story as with extraction. It also works via command. But here the logic a bit more
+ * complicated because cache is able to invalidate related caches by {@link AbstractKeyParam#getEntityType()}.
+ */
+public class Cache {
```

Comment:
It might be a good idea to create an interface to understand API for this class quickly.

Reply:
it's no reason to create an interface coze it means that we can create another realization for the cache